Add comb function to jax._src.scipy.special

Phionx opened the pull request:
Here, we add the comb function to jax.scipy.special. This implementation provides an approximate and efficient calculation, while also handling the exact case.

As in scipy, we also handle the case where the number of repetitions is computed.

Finally, we add a test in the `tests/lax_scipy_special_functions_test.py` file to maintain testing coverage.

This is my first PR to jax. Although I tried to follow the contributing guidelines closely, please let me know if there's anything I missed. 

Thanks!

Resolves: #9709

Phionx committed:
Add comb function to jax._src.scipy.special

Here, we add the comb function to jax.scipy.special.
This implementation provides an approximate and efficient
calculation, while also handling the exact case.

As in scipy, we also handle the case where the number of
repetitions is computed.

Finally, we add a test in the tests/lax_scipy_special_functions_test.py
file to maintain testing coverage.

Resolves: #9709

jakevdp commented on jax/_src/scipy/special.py:
```diff
@@ -1713,3 +1714,47 @@ def bernoulli(n: int) -> Array:
+  if exact:
+    max_divisor = lax.max(k, N - k)
+    min_divisor = lax.min(k, N - k)
+    N_factorial_over_max_factorial = np.prod(np.arange(N, max_divisor, -1))
```
The approach used for `exact` here will not be compatible with `jit` and other transformations, because it relies on creating dynamically-shaped arrays. That said, I'm not sure what to recommend: I haven't seen any good JIT-compatible approach for producing exact integer factorials in JAX.

I'd probably just leave it unimplemented for now.

Phionx replied:
Hm, to clarify, would you suggest that I just delete this altogether or is it still worth leaving it in?

jakevdp replied:
I would delete this and raise a `NotImplementedError`.

jakevdp commented on tests/lax_scipy_special_functions_test.py:
```diff
@@ -195,6 +195,19 @@ def testScipySpecialFunBernoulli(self, n):
+    lax_op = functools.partial(lsp_special.comb, N, k, exact=exact, repetition=repetition)
+    args_maker = lambda: []
+    self._CheckAgainstNumpy(scipy_op, lax_op, args_maker, atol=0, rtol=1E-5)
+
```
Add `self._CompileAndCheck` as well.

jakevdp reviewed (changes requested):
Looks good overall! A few comments below. Also please add the new API to the docs at https://github.com/google/jax/blob/main/docs/jax.scipy.rst#jaxscipyspecial

Phionx committed:
Updating PR #18389 for test on scipy.special.comb

Run `pytest -n auto tests/lax_scipy_special_functions_test.py` to run
the relevant tests.

Phionx committed:
Updating PR #18389 for test on scipy.special.comb

Run `pytest -n auto tests/lax_scipy_special_functions_test.py` to run
the relevant tests.

jakevdp commented on tests/lax_scipy_special_functions_test.py:
```diff
@@ -25,9 +25,12 @@
 
 from jax import config
 config.parse_flags_with_absl()
+config.update("jax_enable_x64", True)
```
We should not enable x64 here; this will override the environment settings that we use to run tests both with and without x64 enabled.

jakevdp commented on jax/_src/scipy/special.py:
```diff
@@ -1719,3 +1720,47 @@ def bernoulli(n: int) -> Array:
+    \end{equation}
+  
+  Where we use the Gamma function. 
+  """
```
You can remove this entire docstring, it will be overwritten by the `_wraps` decorator above.

Phionx committed:
Merge branch 'main' of https://www.github.com/google/jax into adding-scipy-special-comb

Phionx committed:
Removed docstring and exact implementation,
removed x64  precision from test

Phionx committed:
Merge branch 'main' of https://www.github.com/google/jax into adding-scipy-special-comb

Phionx commented:
Hi @jakevdp,

Thanks for the ping! I've made changes based on your most recent suggestions. 

Please let me know if this looks good to you or if there are any other changes for me to add.

Thanks for your review!

jakevdp reviewed (changes requested):
Looking good! A few more comments and we can get this merged!

jakevdp commented on jax/_src/scipy/special.py:
```diff
@@ -1901,3 +1902,22 @@ def hyp1f1(a, b, x):
+
+@_wraps(osp_special.comb, module='scipy.special')
+def comb(N: Array, k: Array, exact: bool =False, repetition: bool =False):
+  #  Calculates N choose k, the number of ways to choose k items from N items, with and without repetition.
```
We should call `promote_dtypes_inexact` here, because `gammaln` expects inexact inputs.

Phionx replied:
I believe `scipy.special.comb` allows for non-integer `N` and `k`, so we may wish to do the same. 

Also, it seems `promote_args_inexact` already uses `promote_dtypes_inexact`. Do I still need to use `promote_dtypes_inexact` anywhere, if I use `promote_args_inexact`?

jakevdp replied:
`promote_args_inexact` will call `promote_dtypes_inexact` and also broadcast the arguments to a common shape, which I don't think we want here.

jakevdp commented on jax/_src/scipy/special.py:
```diff
@@ -1901,3 +1902,22 @@ def hyp1f1(a, b, x):
 )
+
+@_wraps(osp_special.comb, module='scipy.special')
+def comb(N: Array, k: Array, exact: bool =False, repetition: bool =False):
```
Input arrays should be annotated `ArrayLike`, and we should annotate the return value as ` -> Array`.

jakevdp commented on jax/_src/scipy/special.py:
```diff
@@ -1901,3 +1902,22 @@ def hyp1f1(a, b, x):
+    # max_divisor = lax.max(k, N - k)
+    # min_divisor = lax.min(k, N - k)
+    # N_factorial_over_max_factorial = np.prod(np.arange(N, max_divisor, -1))
+    # return lax.div(N_factorial_over_max_factorial, np.prod(np.arange(1, min_divisor + 1)))
```
I'd remove the commented code here.

jakevdp commented on jax/_src/scipy/special.py:
```diff
@@ -1901,3 +1902,22 @@ def hyp1f1(a, b, x):
+  one = _constant_like(N, 1)
+  N_plus_1 = lax.add(N,one)
+  k_plus_1 = lax.add(k,one)
+  return lax.exp(lax.sub(gammaln(N_plus_1),lax.add(gammaln(k_plus_1), gammaln(lax.sub(N_plus_1,k)))))
```
Also, we should think about what the expected return value is when `k > N`. What does `scipy.special.comb` do in this case?

CompRhys commented:
Can I bump this? Am looking to migrate some bezier curve code from scipy to jax to see if jax can be used to fit composite bezier curves efficiently
